osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,51 +20,165 @@ public class Speed : OsuStrainSkill
+
+                        else
+                        {
+                            if (islandSize > 6)
+                                islandSize = 6;
+
+                            if (Previous[i - 1].BaseObject is Slider) // bpm change is into slider, this is easy acc window
+                                effectiveRatio *= 0.25;
+
+                            if (Previous[i].BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
+                                effectiveRatio *= 0.5;
+
+                            if (previousIslandSize == islandSize) // repeated island size (ex: triplet -> triplet)
+                                effectiveRatio *= 0.25;
+
+                            rhythmComplexitySum += effectiveRatio;
+
+                            previousIslandSize = islandSize; // log the last island size.
+
+                            if (prevDelta * 1.25 < currDelta) // we're slowing down, stop counting
+                                firstDeltaSwitch = false; // if we're speeding up, this stays true and  we keep counting island size.
+
+                            islandSize = 0;
+                        }
+                    }
+                    else if (prevDelta > 1.25 * currDelta) // we want to be speeding up.
+                    {
+                        // Begin counting island until we change speed again.
+                        firstDeltaSwitch = true;
+                        islandSize = 0;
+                    }
+                }
+            }
+
+            return Math.Sqrt(4 + rhythmComplexitySum * rhythm_multiplier) / 2; //produces multiplier that can be applied to strain. range [1, infinity)
+        }
+
+        private double tapStrainOf(DifficultyHitObject current, double speedBonus)
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+
+            return speedBonus / osuCurrObj.StrainTime;
+        }
+
+        private double movementStrainOf(DifficultyHitObject current, double speedBonus)
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+
+            double distance = Math.Min(single_spacing_threshold, osuCurrObj.TravelDistance + osuCurrObj.JumpDistance);
 
             double angleBonus = 1.0;
 
-            if (osuCurrent.Angle != null && osuCurrent.Angle.Value < angle_bonus_begin)
+            if (osuCurrObj.Angle != null)
             {
-                angleBonus = 1 + Math.Pow(Math.Sin(1.5 * (angle_bonus_begin - osuCurrent.Angle.Value)), 2) / 3.57;
+                double angle = osuCurrObj.Angle.Value;
+
+                if (angle < pi_over_2)
+                    angleBonus = 1.25;
+                else if (angle < angle_bonus_begin)
+                    angleBonus = 1 + Math.Pow(Math.Sin(1.5 * (angle_bonus_begin - angle)), 2) / 4;
```

Comment:
These seem to have changed ever so slightly from what the previous angle bonus calculation did, was there any particular reason for those tweaks?

Reply:
In this change, angleBuff is removed from the "Tap" portion of the strain. Because of this, I thought it would be a good idea to refactor the ugliness of the old angle algorithm. It still follows the same mathematics and angles, but is implemented cleaner.